label-application-rule-config-service-impl/src/main/java/org/hypertrace/label/application/rule/config/service/LabelApplicationRuleConfigServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+      responseObserver.onNext(
+          CreateLabelApplicationRuleResponse.newBuilder()
+              .setLabelApplicationRule(createdLabelApplicationRule)
+              .build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void getLabelApplicationRule(
+      GetLabelApplicationRuleRequest request,
+      StreamObserver<GetLabelApplicationRuleResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      this.requestValidator.validateOrThrow(requestContext, request);
+      String ruleId = request.getId();
+      Optional<LabelApplicationRule> optionalLabelApplicationRule =
+          this.labelApplicationRuleStore.getObject(requestContext, ruleId);
+      if (optionalLabelApplicationRule.isEmpty()) {
+        responseObserver.onError(new StatusRuntimeException(Status.NOT_FOUND));
+        return;
+      }
+      responseObserver.onNext(
+          GetLabelApplicationRuleResponse.newBuilder()
+              .setLabelApplicationRule(optionalLabelApplicationRule.get())
+              .build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void getLabelApplicationRules(
+      GetLabelApplicationRulesRequest request,
+      StreamObserver<GetLabelApplicationRulesResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      this.requestValidator.validateOrThrow(requestContext, request);
+      List<LabelApplicationRule> labelApplicationRules =
+          this.labelApplicationRuleStore.getAllObjects(requestContext);
+      responseObserver.onNext(
+          GetLabelApplicationRulesResponse.newBuilder()
+              .addAllLabelApplicationRules(labelApplicationRules)
+              .build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void updateLabelApplicationRule(
+      UpdateLabelApplicationRuleRequest request,
+      StreamObserver<UpdateLabelApplicationRuleResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      this.requestValidator.validateOrThrow(requestContext, request);
+      String ruleId = request.getId();
+      LabelApplicationRule labelApplicationRule =
+          LabelApplicationRule.newBuilder().setId(ruleId).setData(request.getData()).build();
+      if (this.labelApplicationRuleStore.getObject(requestContext, ruleId).isEmpty()) {
+        responseObserver.onError(new StatusRuntimeException(Status.NOT_FOUND));
+        return;
+      }
+      LabelApplicationRule updateLabelApplicationRule =
+          this.labelApplicationRuleStore.upsertObject(requestContext, labelApplicationRule);
```

Comment:
Suggestion (avoids extra return and removes assumption about structure of a rule):

```java
      LabelApplicationRule existingRule =
          this.labelApplicationRuleStore
              .getObject(requestContext, request.getId())
              .orElseThrow(Status.NOT_FOUND::asException);

      LabelApplicationRule updateLabelApplicationRule =
          existingRule.toBuilder().setData(request.getData()).build();
```

What I mean by the assumption is that if existingRule contained other fields that weren't part of the update flow (imagine something like creation timestamp), they'll be preserved, which would need to be explicit when creating a brand new rule object from scratch in the current approach.